### PR TITLE
Support building with MSVC

### PR DESCRIPTION
**Pull request by DomClark**
See also LMMS/lmms#6561.

This pull request mostly consists of routine transformations from POSIX APIs to the C++ standard library, along with `_alloca` alternatives for variable-length stack arrays. In order to keep things easy to review, and maybe pull in other upstream changes, I haven't tried to improve the code by using things such as RAII lock guards (except where required, i.e. in conjunction with a condition variable). As an exception, I took the opportunity to support Unicode paths in a few places by using wide character functions on Windows.

### CODE REVIEW

File: src/DSP/AnalogFilter.cpp [hunk 1]
```diff
@@ -391,7 +391,11 @@ void AnalogFilter::filterout(float *smp)
 
     if(needsinterpolation) {
         //Merge Filter at old coeff with new coeff
+#ifdef _MSC_VER
+        const auto ismp = static_cast<float*>(_alloca(buffersize * sizeof(float)));
+#else
         float ismp[buffersize];
+#endif
```
Comment (dan-giddins) [inline src/DSP/AnalogFilter.cpp:398]: Couldn't you just do something like `auto ismp = new float[buffersize];` for both GCC and MSVC? I am pretty much a novice at C++ so I would be interested to see why something like this wouldn't work. Are we on C++11?

Comment (DomClark) [reply src/DSP/AnalogFilter.cpp:398]: That would work from a C++ perspective, but this is realtime code, so heap allocations (using `new`, `make_unique`, etc.) aren't appropriate. There's no bound on how long the memory allocation will take, so you can't guarantee the audio will be generated in time.

Standard C++ doesn't support variable-length stack arrays, which is why this is done in different ways for different compilers. GCC supports such arrays as a non-standard language extension, and MSVC supports the non-standard function `_alloca` to do the same thing. (GCC also supports `alloca` from what I understand, but I'm following the technique we've used in the past in LMMS for compatibility.) Ideally we would stick to standard C++, which would probably involve doing a heap allocation ahead of time on a non-realtime thread. But I'm trying to change as little as possible in this PR as this repo mostly contains third-party code.

Comment (dan-giddins) [reply src/DSP/AnalogFilter.cpp:398]: Ahh I see, thank you for explaining! I hope to help contribute to getting the MSVC build working when I get a chance so this is good to know!

File: src/Nio/ZynSema.h [hunk 2]
```diff
@@ -12,45 +12,38 @@ class ZynSema
...
-        return pthread_mutex_init (&_mutex, 0) || pthread_cond_init (&_cond, 0);
+        return 0;
     }
 
     int post (void)
     {
-        pthread_mutex_lock (&_mutex);
-        if (++_count == 1) pthread_cond_signal (&_cond);
-        pthread_mutex_unlock (&_mutex);
+        _mutex.lock();
```
Comment (PhysSong) [inline src/Nio/ZynSema.h:23]: Why this function uses manual lock-unlock calls whereas `wait` uses `std::unique_lock`?

Comment (DomClark) [reply src/Nio/ZynSema.h:23]: My aim is not to improve the code by switching to RAII lock guards; I'm only trying to get the existing code to build with MSVC. I chose to keep manual lock-unlock calls as then each new line clearly corresponds to one old line. This should hopefully make resolving future conflicts with any upstream changes easier, as well as make this PR easier to review. However, `std::condition_variable::wait` requires a `std::unique_lock` as an argument, so I had to make an exception for that one function.

**Review by PhysSong — APPROVED** [review]

**Comment by JohannesLorenz** [comment]
:+1:

Many parts of this PR might not be applied to current zyn's master (e.g. some variables do not occur there, and we meanwhile also have the `STACKALLOC` macro for stack allocation), but a few could (e.g. the `mkdir` changes). Are you going to push those parts that can be applied to zyn's master, too?

**Comment by DomClark** [comment]
> Are you going to push those parts that can be applied to zyn's master, too?

This is something I definitely plan to do eventually, especially since we plan on updating our Zyn version and presumably still want to be able to build everything with MSVC. Right now, this is mostly for personal convenience as I'm working on our remote plugin IPC interface and want to be able to test my changes with Zyn more easily.

**Comment by PhysSong** [comment]
In a second look, I noticed that `ThreadShims.h` is outside this repository. Is it safe to do such a thing?

**Comment by PhysSong** [comment]
Oh, We already do that for `IoHelper.h`.

**Comment by DomClark** [comment]
Yeah, it's not ideal, but I needed it in the LMMS repository too, and didn't want to duplicate it. As you say, we already do the same for `IoHelper.h`, and hopefully we can get rid of it soon once we can use a MinGW version with built-in support.

**Review by PhysSong — APPROVED** [review]

**Pull request merged**